### DISCUSSION TITLE
Running shell apps from textual

### DISCUSSION THREAD

**Discussion by artur-shaik** [Textualize/textual#165] [Q&A]
Any hints on execute shell app, like vim, from key event method?

Now using `os.system` or `subprocess` or `asyncio.create_subprocess_shell`, when I close vim, textual app stops, I even cannot close it with `Ctrl-c`

**Comment by nraghuveer** [comment]
How are you closing vim? ":qa"?

**Reply by artur-shaik** [reply]
Yes, with `:qa`.

**Comment by mwalkerr** [comment]
@artur-shaik I'm also interested in doing something like this. Did you ever figure out how to get it to work?

**Reply by mwalkerr** [reply]
I think we need to use a PTY (https://docs.python.org/3/library/pty.html) to interact with other terminal applications that generally produce output directly for a terminal

**Reply by artur-shaik** [reply]
Not actually. I decided to start one more terminal window with `vim` executed there.

**Comment by geoff-va** [comment]
I'd love to get this working as well.  A way to more or less seamlessly drop into vim and do some editing and then pop back into the application would be great. I don't think I can start it in another window since the application is being run on another server in an ssh session (unless someone knows how to do that?).

I've found that dropping directly into vim "works", but I've got to press all the keys twice for them to register.

If I call `driver.stop_application_mode()`, then vim properly works as expected.

The issue is then resuming application mode, it seems. I've dug into `driver.start_application_mode` and tried to pull out the relevant components that seem to be affecting the way the terminal is setup, but it becomes unresponsive at that point. I'm not sure if it's because the asyncio loop isn't properly running anymore or what. 

If anyone has any thoughts I'd love to hear them!

**Reply by napisani** [reply]
I'm curious if you all made any progress with this? I'm building a TUI using textual right now and hoping implement something similar.
if you have any code pointers or suggestions it would be super helpful. I'll be sure to share too if i come up with anything.

**Comment by antaenc** [comment]
I don't embed into the Textual layout, I launch a new terminal with vi in it to edit code. :wq then saves and exits and control is given back to my app. This is the CSS branch, still learning this I'm sure there are other ways but this is what I have working atm. Note, I'm only testing for gnome atm as that is what I'm using

```
    def action_edit(self):

        if self.current_file:
            de = getenv('XDG_CURRENT_DESKTOP', '')
            if de and 'GNOME' in de:
                # Wait for vi to exit
                with subprocess.Popen(["/usr/bin/gnome-terminal",
                                       "--geometry=120x40",
                                       "--wait",
                                       "--",
                                       "vi",
                                       f'{str(self.current_file)}']) as vi_exec:
                    vi_exec.wait()

                syntax = Syntax.from_path(
                        self.current_file,
                        line_numbers=True,
                        word_wrap=False,
                        indent_guides=False,
                        theme="dracula"
                    )
                
                self.query_one("#code", Static).update(syntax)
```

**Reply by WaylonWalker** [reply]
I've done something very similar, but using tmux rather than a new shell with success.  Feels a bit hacky and wouldn't work for everyone, but is a solution for a personal project if you use tmux.

**Comment by davep** [comment]
In case it's of interest to anyone, there is now an issue for this: https://github.com/Textualize/textual/issues/1093

**Comment by bdhammel** [comment]
This strategy seems to be working okay for me

```
        self._driver.stop_application_mode()
        initial_text = 'test'
        try:
            with tempfile.NamedTemporaryFile(suffix=".txt") as tf:
                tf.write(bytes(initial_text, 'UTF-8'))
                tf.flush()
                # Need to create a separate backup copy
                # If we don't, the edited text will not be saved into the current file
                subprocess.call(['vim', '+set backupcopy=yes', tf.name])
                tf.seek(0)
                input_ = tf.read().decode("utf-8")
        finally:
            self.refresh()
            self._driver.start_application_mode()

```

**Reply by bagaag** [reply]
Thanks - this works for me, too. It flashes the terminal screen "behind" the TUI when switching between the TUI and the launched app, but it works seamlessly otherwise.

**Reply by miraculixx** [reply]
This works great. ~~If've adopted your solution by adding a~~ Actually, textual  now provides a [`suspend()` method ](https://textual.textualize.io/guide/app/#suspending) to the app class, as a context manager. So now it's possible to write code like:

``` python
    def action_foobar(*args):
       with self.suspend():
         # all processing goes here
         subprocess.call(...)       
```

The actual `suspend()` method does a bit more work, but in effect it does this:

``` python
    class MyApp(App):
         ...
      @contextmanager
      def suspend(self):
        self._driver.stop_application_mode()
        yield
        self._driver.start_application_mode()
        self.refresh()
```

**Reply by TomJGooding** [reply]
Perhaps this code is doing something different, but in case you missed it Textual actually added App.suspend in v0.48.0:

https://textual.textualize.io/guide/app/#suspending

**Reply by davep** [reply]
The above version of `suspend` is different from the one in Textual itself, in that [it would seem to be less safe](https://github.com/Textualize/textual/blob/5b3a2e86269d805a11f9d03dccd3ca757087ddd2/src/textual/app.py#L3456-L3504).

**Reply by miraculixx** [reply]
Thanks, didn't realize textual provided a suspend method. Have updated my reply above to reflect that.
